backend/routers/buildings.js

```diff
@@ -1,68 +1,77 @@
 const { Router } = require('express');
 const createError = require('http-errors');
-const { BuildingModel } = require('../models');
+const { BuildingModel, ResourceModel } = require('../models');
 
 const router = Router();
 
 const {
   townhallRule, farmRule, mineRule, academyRule, foxRule,
 } = require('../rules');
 
+const typeToRules = {
+  Townhall: townhallRule,
+  Academy: academyRule,
+  Farm: farmRule,
+  Mine: mineRule,
+};
+
 async function getBuildings(req, res, next) {
   try {
-    const buildings = await BuildingModel.find({}, '-_id');
+    const buildings = await BuildingModel.find({ owner: 1 }, '-_id');
     res.send({ buildings });
   } catch (error) {
     next(error);
   }
 }
 
 async function createBuilding(req, res, next) {
-  const type = req.params.buildingType;
-  const buildings = await BuildingModel.find({});
-  const index = buildings.length ? buildings.length + 1 : 1;
+  const { buildingType } = req.params;
   try {
-    const result = await BuildingModel.create({
-      id: index,
-      type,
-      level: 1,
+    if (!buildingType) {
+      throw createError(400, 'Type is required');
+    }
+    if (buildingType !== 'Mine' && buildingType !== 'Farm') {
+      throw createError(400, 'Wrong type');
+    }
+
+    const moneyRequired = parseInt(typeToRules[buildingType].constructionCost, 10);
+    const resourceOfUser = await ResourceModel.find({ owner: 1 });
+    const [moneyInHand] = resourceOfUser
+      .filter(({ type }) => (type === 'gold'))
+      .map(({ amount }) => amount);
+    if (moneyRequired > moneyInHand) {
+      throw createError(400, 'You don\'t have enough money');
+    }
+
+    const minusGold = await ResourceModel.findOneAndUpdate(
+      { owner: 1, type: 'gold' },
+      { $inc: { amount: -{ moneyRequired } } },
+      { new: true, fields: '-_id' },
+    ).exec();
```

Comment:
Resource model should provide a method like `reduceMoney(amount)` to let other module call.